docs/notebooks/Hidden Linear Function Problem.ipynb

```diff
@@ -0,0 +1,522 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Hidden Linear Function Problem\n",
+    "\n",
+    "In this notebook we consider a problem from the paper \"[Quantum advantage with shallow circuits](https://arxiv.org/pdf/1704.00690.pdf)\" and build a quantum circuit, which solves it, in Cirq. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "### The problem\n",
+    "\n",
+    "In this problem we consider a quadratic form of a binary vector and with binary coefficients (but additions and multiplications are evaluated modulo 4). Then we restrict this quadratic form, i.e. we allow to use only certain binary vectors as input. It turns out that under this restriction this quadratic form is equivalent to a linear function, i.e. it just evaluates dot product of input vector and certain scalar vector. Task is to find this scalar vector.\n",
```

Comment:
Link quadratic form (https://en.wikipedia.org/wiki/Quadratic_form) and binary vector: https://en.wikipedia.org/wiki/Bit_array